kroxylicious/src/main/java/io/kroxylicious/proxy/internal/DefaultFilterContext.java

```diff
@@ -35,7 +38,7 @@ class DefaultFilterContext implements KrpcFilterContext {
 
     private static final Logger LOGGER = LoggerFactory.getLogger(DefaultFilterContext.class);
 
-    private final DecodedFrame<?, ?> decodedFrame;
+    private DecodedFrame<?, ?> decodedFrame;
```

Comment:
Instead of making this mutable we could construct the new Frame inside `forwardResponse` and call `fireChannelRead` without updating the field. Maybe contexts should mark themselves as being "finalised" to prevent filters making multiple forwarding calls to the same response?

Reply:
I am very tempted to say the context should be immutable, constructing a new frame inside `forwardResponse` seems like the right choice to me. However can't quite wrap my head around the implications of making multiple calls to `forward`.

Reply:
yeah, making that change around the mutability now, we don't need to update the field.

Reply:
I think I saw a TODO about preventing repeat invocations, yeah theres ` // TODO check we've not forwarded it already` in both of the forward methods.